daemon/containerd/cache.go

```diff
@@ -14,80 +15,230 @@ import (
 // MakeImageCache creates a stateful image cache.
 func (i *ImageService) MakeImageCache(ctx context.Context, cacheFrom []string) (builder.ImageCache, error) {
 	images := []*image.Image{}
-	for _, c := range cacheFrom {
-		im, err := i.GetImage(ctx, c, imagetype.GetImageOpts{})
-		if err != nil {
-			return nil, err
+	if len(cacheFrom) == 0 {
+		return &localCache{
+			imageService: i,
+		}, nil
+	} else {
+		for _, c := range cacheFrom {
+			h, err := i.ImageHistory(ctx, c)
+			if err != nil {
+				continue
+			}
+			for _, hi := range h {
+				if hi.ID != "<missing>" {
+					im, err := i.GetImage(ctx, hi.ID, imagetype.GetImageOpts{})
+					if err != nil {
+						return nil, err
+					}
+					images = append(images, im)
+				}
+			}
 		}
-		images = append(images, im)
 	}
-	return &imageCache{images: images, c: i}, nil
+
+	return &imageCache{
+		lc: &localCache{
+			imageService: i,
+		},
+		images:       images,
+		imageService: i,
+	}, nil
 }
 
-type imageCache struct {
-	images []*image.Image
-	c      *ImageService
+type localCache struct {
+	imageService *ImageService
 }
 
-func (ic *imageCache) GetCache(parentID string, cfg *container.Config) (imageID string, err error) {
+func (ic *localCache) GetCache(parentID string, cfg *container.Config) (imageID string, err error) {
 	ctx := context.TODO()
 
+	var children []image.ID
+
+	// FROM scratch
 	if parentID == "" {
-		// TODO handle "parentless" image cache lookups ("FROM scratch")
-		return "", nil
+		imgs, err := ic.imageService.Images(ctx, types.ImageListOptions{
+			All: true,
+		})
+		if err != nil {
+			return "", err
+		}
+		for _, img := range imgs {
+			children = append(children, image.ID(img.ID))
+		}
+	} else {
```

Comment:
I think this is incorrect. Any image is not a suitable cache for the `FROM scratch` image. This needs to be only the intermediate images that were built `FROM scratch`.

Not sure what's the best way to detect them though - because it looks like containerd doesn't store labels with empty values, so we can't really just check if `﻿org.mobyproject.image.parent` label exists but is empty.

Reply:
In the original case, if it's a `FROM scratch` we take the whole image map https://github.com/moby/moby/blob/master/image/cache/cache.go#L239

Reply:
But they're also filtered with `img.Parent == ""`: https://github.com/moby/moby/blob/777e9f271095685543f30df0ff7a12397676f938/image/cache/cache.go#L243

Reply:
Right, let me add that, see if anything breaks

Reply:
I don't think it's enough - the Parent will be empty for pulled/buildkit images so they will be considered a valid cache too.